src/java.desktop/macosx/native/libosxapp/JNIUtilities.h

```diff
@@ -206,4 +206,43 @@
        return y; \
     };
 
+/* Create a pool and initiate a try block to catch any exception */
+#define JNI_COCOA_ENTER(env) \
+ NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init]; \
+ @try {
+
+/* Don't allow NSExceptions to escape to Java.
+ * If there is a Java exception that has been thrown that should escape.
+ * And ensure we drain the auto-release pool.
+ */
+#define JNI_COCOA_EXIT(env) \
+ } \
+ @catch (NSException *e) { \
+     NSLog(@"%@", [e callStackSymbols]); \
+ } \
+ @finally { \
+    [pool drain]; \
+ };
+
+/* Same as above but adds a clean up action.
+ * Requires that whatever is being cleaned up is in scope.
+ */
+#define JNI_COCOA_EXIT_WITH_ACTION(env, action) \
+ } \
+ @catch (NSException *e) { \
+     { action; }; \
+     NSLog(@"%@", [e callStackSymbols]); \
+ } \
+ @finally { \
+    [pool drain]; \
+ };
+
+#define JNI_COCOA_THROW_OOME(env, msg) \
+    JNU_ThrowOutOfMemoryError(env, msg); \
+    [NSException raise:@"Java Exception" reason:@"Java OutOfMemoryException" userInfo:nil]
+
+#define JNI_COCOA_THROW_RUNTIME_EXCEPTION(env, msg) \
+    JNU_ThrowByName(env, "java/lang/RuntimeException", msg); \
```

Comment:
Can we guarantee that before the next java code executed later we clean this exception? 

BTW I still think the exceptions wrapping will be a better thing. If it will be caught by our catch block like in JNI_COCOA_EXIT we can unwrap it, otherwise, the java code could be executed safely since we clear an exception during wrapping.

Reply:
Since the next line throws NSException everything before this change and after this change should cause control to reach the COCOA_EXIT macro.  I am not aware of anything in between. So this should be OK. The previous JNF code (as I discussed in some earlier review / thread) seems to be *internally* responsible for not doing things properly as I had found that its exception handling code actually made calls that aren't allowed with the exception pending. And this should not happen any more.

Reply:
> Since the next line throws NSException everything before this change and after this change should cause control to reach the COCOA_EXIT macro.

It does not, the JNI_COCOA_EXIT macro is executed on a different thread than JNI_COCOA_THROW_RUNTIME_EXCEPTION (in both times where this macro is used)

Reply:
You are correct. This is running in a performOnMainThreadWaiting:NO block.
So the JNI call likely already returned to Java before that block gets run on the AppKit thread.
But this seems to be the same as before and there was probably never any point to raising the JDK exception
I think in this case maybe what I should is update the macro to skip throwing the Java Exception if
it is being raised on the AppKit thread. No caller that I can see is expecting a RuntimeException for these failed actions so this seems to be very much an implementation decision. I am not even sure the code should be doing this. Rightly or wrongly GraphicsDevice.setFullScreenWindow(Window) declares no exceptions or possibility of failure.

Reply:
It is always better to wrap the java exception and always clear after each callJavaMethod, so the next callJavaMethod will not be called when the java exception was raised. At the same time in the place where we catch the nsexception we can check does it have a java exception or not -> so we can unwrap and raise again the java exception just before return to the java code, or such exception in the NSApplication global catch block(including currently missing NSApplication#reportException).

Reply:
It is one way of doing it but it is much more heavyweight as it requires a carrier for the exception
and so goes against the approach here which is to be very lightweight in the sense of avoiding
creating classes - which so far as been the case.
And there is no need to wrap it. If its there, it will be just be passed on up to Java without any
work on our part.
And as for clearing, that's what I was doing before and that made sense if the intention is to carry on.
Since now we are always raising NSException there  that next Java call shouldn't happen as we always
should jump out.